File: plugins/wasm-go/pkg/wrapper/cluster_wrapper.go
// Copyright (c) 2022 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package wrapper

import (
	"fmt"
	"strings"

	"github.com/higress-group/proxy-wasm-go-sdk/proxywasm"
)

type Cluster interface {
	ClusterName() string
	HostName() string
}

type RouteCluster struct {
	Host string
}

func (c RouteCluster) ClusterName() string {
	routeName, err := proxywasm.GetProperty([]string{"cluster_name"})
	if err != nil {
		proxywasm.LogErrorf("get route cluster failed, err:%v", err)
	}
	return string(routeName)
}

func (c RouteCluster) HostName() string {
	if c.Host != "" {
		return c.Host
	}
	return GetRequestHost()
}

type K8sCluster struct {
	ServiceName string
	Namespace   string
	Port        int64
	Version     string
	Host        string
}

func (c K8sCluster) ClusterName() string {
	namespace := "default"
	if c.Namespace != "" {
		namespace = c.Namespace
	}
	return fmt.Sprintf("outbound|%d|%s|%s.%s.svc.cluster.local",
		c.Port, c.Version, c.ServiceName, namespace)
}

func (c K8sCluster) HostName() string {
	if c.Host != "" {
		return c.Host
	}
	return fmt.Sprintf("%s.%s.svc.cluster.local", c.ServiceName, c.Namespace)
}

type NacosCluster struct {
	ServiceName string
	// use DEFAULT-GROUP by default
	Group       string
	NamespaceID string
	Port        int64
	// set true if use edas/sae registry
	IsExtRegistry bool
	Version       string
	Host          string
}

func (c NacosCluster) ClusterName() string {
	group := "DEFAULT-GROUP"
	if c.Group != "" {
		group = strings.ReplaceAll(c.Group, "_", "-")
	}
	tail := "nacos"
	if c.IsExtRegistry {
		tail += "-ext"
	}
	return fmt.Sprintf("outbound|%d|%s|%s.%s.%s.%s",
		c.Port, c.Version, c.ServiceName, group, c.NamespaceID, tail)
}

func (c NacosCluster) HostName() string {
	if c.Host != "" {
		return c.Host
	}
	return c.ServiceName
}

type StaticIpCluster struct {
	ServiceName string
	Port        int64
	Host        string
}

func (c StaticIpCluster) ClusterName() string {
	return fmt.Sprintf("outbound|%d||%s.static", c.Port, c.ServiceName)
}

func (c StaticIpCluster) HostName() string {
	if c.Host != "" {
		return c.Host
	}
	return c.ServiceName
}

type DnsCluster struct {
	ServiceName string
	Domain      string
	Port        int64
}

func (c DnsCluster) ClusterName() string {
	return fmt.Sprintf("outbound|%d||%s.dns", c.Port, c.ServiceName)
}

func (c DnsCluster) HostName() string {
	return c.Domain
}

type ConsulCluster struct {
	ServiceName string
	Datacenter  string
	Port        int64
	Host        string
}

func (c ConsulCluster) ClusterName() string {
	tail := "consul"
	return fmt.Sprintf("outbound|%d||%s.%s.%s",
		c.Port, c.ServiceName, c.Datacenter, tail)
}

func (c ConsulCluster) HostName() string {
	if c.Host != "" {
		return c.Host
	}
	return c.ServiceName
}

type FQDNCluster struct {
	FQDN string
	Host string
	Port int64
}

func (c FQDNCluster) ClusterName() string {
	return fmt.Sprintf("outbound|%d||%s", c.Port, c.FQDN)
}

func (c FQDNCluster) HostName() string {
	if c.Host != "" {
		return c.Host
	}
	return c.FQDN
}

File: plugins/wasm-assemblyscript/assembly/cluster_wrapper.ts
import {
  log,
  LogLevelValues,
  get_property,
  WasmResultValues,
} from "@higress/proxy-wasm-assemblyscript-sdk/assembly";
import { getRequestHost } from "./request_wrapper";
  
export abstract class Cluster {
  abstract clusterName(): string;
  abstract hostName(): string;
}
  
export class RouteCluster extends Cluster {
  host: string;
  constructor(host: string = "") {
    super();
    this.host = host;
  }
  
  clusterName(): string {
    let result = get_property("cluster_name");
    if (result.status != WasmResultValues.Ok) {
      log(LogLevelValues.error, "get route cluster failed");
      return "";
    }
    return String.UTF8.decode(result.returnValue);
  }
  
  hostName(): string {
    if (this.host != "") {
      return this.host;
    }
    return getRequestHost();
  }
}
  
export class K8sCluster extends Cluster {
  serviceName: string;
  namespace: string;
  port: i64;
  version: string;
  host: string;

  constructor(
    serviceName: string,
    namespace: string,
    port: i64,
    version: string = "",
    host: string = ""
  ) {
    super();
    this.serviceName = serviceName;
    this.namespace = namespace;
    this.port = port;
    this.version = version;
    this.host = host;
  }

  clusterName(): string {
    let namespace = this.namespace != "" ? this.namespace : "default";
    return `outbound|${this.port}|${this.version}|${this.serviceName}.${namespace}.svc.cluster.local`;
  }

  hostName(): string {
    if (this.host != "") {
      return this.host;
    }
    return `${this.serviceName}.${this.namespace}.svc.cluster.local`;
  }
}

export class NacosCluster extends Cluster {
  serviceName: string;
  group: string;
  namespaceID: string;
  port: i64;
  isExtRegistry: boolean;
  version: string;
  host: string;

  constructor(
    serviceName: string,
    namespaceID: string,
    port: i64,
    // use DEFAULT-GROUP by default
    group: string = "DEFAULT-GROUP",
    // set true if use edas/sae registry
    isExtRegistry: boolean = false,
    version: string = "",
    host: string = ""
  ) {
    super();
    this.serviceName = serviceName;
    this.group = group.replace("_", "-");
    this.namespaceID = namespaceID;
    this.port = port;
    this.isExtRegistry = isExtRegistry;
    this.version = version;
    this.host = host;
  }

  clusterName(): string {
    let tail = "nacos" + (this.isExtRegistry ? "-ext" : "");
    return `outbound|${this.port}|${this.version}|${this.serviceName}.${this.group}.${this.namespaceID}.${tail}`;
  }

  hostName(): string {
    if (this.host != "") {
      return this.host;
    }
    return this.serviceName;
  }
}

export class StaticIpCluster extends Cluster {
  serviceName: string;
  port: i64;
  host: string;

  constructor(serviceName: string, port: i64, host: string = "") {
    super()
    this.serviceName = serviceName;
    this.port = port;
    this.host = host;
  }

  clusterName(): string {
    return `outbound|${this.port}||${this.serviceName}.static`;
  }

  hostName(): string {
    if (this.host != "") {
      return this.host;
    }
    return this.serviceName;
  }
}

export class DnsCluster extends Cluster {
  serviceName: string;
  domain: string;
  port: i64;

  constructor(serviceName: string, domain: string, port: i64) {
    super();
    this.serviceName = serviceName;
    this.domain = domain;
    this.port = port;
  }

  clusterName(): string {
    return `outbound|${this.port}||${this.serviceName}.dns`;
  }

  hostName(): string {
    return this.domain;
  }
}

export class ConsulCluster extends Cluster {
  serviceName: string;
  datacenter: string;
  port: i64;
  host: string;

  constructor(
    serviceName: string,
    datacenter: string,
    port: i64,
    host: string = ""
  ) {
    super();
    this.serviceName = serviceName;
    this.datacenter = datacenter;
    this.port = port;
    this.host = host;
  }

  clusterName(): string {
    return `outbound|${this.port}||${this.serviceName}.${this.datacenter}.consul`;
  }

  hostName(): string {
    if (this.host != "") {
      return this.host;
    }
    return this.serviceName;
  }
}

export class FQDNCluster extends Cluster {
  fqdn: string;
  host: string;
  port: i64;

  constructor(fqdn: string, port: i64, host: string = "") {
    super();
    this.fqdn = fqdn;
    this.host = host;
    this.port = port;
  }

  clusterName(): string {
    return `outbound|${this.port}||${this.fqdn}`;
  }

  hostName(): string {
    if (this.host != "") {
      return this.host;
    }
    return this.fqdn;
  }
}
